src/System.Net.Http/src/System/Net/Http/HttpContent.cs

```diff
@@ -115,30 +112,19 @@ public Task<string> ReadAsStringAsync()
                 // then check for a byte-order-mark (BOM) in the data to figure out the encoding.
                 if (encoding == null)
                 {
-                    byte[] preamble;
-                    foreach (Encoding testEncoding in s_encodingsWithBom)
-                    {
-                        preamble = testEncoding.GetPreamble();
-                        if (ByteArrayHasPrefix(data, dataLength, preamble))
-                        {
-                            encoding = testEncoding;
-                            bomLength = preamble.Length;
-                            break;
-                        }
-                    }
+                    TryDetectEncoding(data, dataLength, ref encoding, ref bomLength);
                 }
 
-                // Use the default encoding if we couldn't detect one.
-                encoding = encoding ?? DefaultStringEncoding;
-
-                // BOM characters may be present even if a charset was specified.
-                if (bomLength == -1)
+                if (encoding == null)
                 {
-                    byte[] preamble = encoding.GetPreamble();
-                    if (ByteArrayHasPrefix(data, dataLength, preamble))
-                        bomLength = preamble.Length;
-                    else
-                        bomLength = 0;
+                    // Use the default encoding if we couldn't detect one.
+                    encoding = DefaultStringEncoding;
+                    bomLength = 0;
+                }
+                else if (bomLength == -1)
```

Comment:
With the way this has been restructured, is this ```else if``` necessary?  I'm wondering if you can just move the body that sets bomLength up into the try/catch earlier, e.g.
```C#
try
{
    encoding = Encoding.GetEncoding(innerThis.Headers.ContentType.CharSet);
    bomLength = GetPreambleLength(data, dataLength, encoding);
}
```
To me that makes all of this logic more clear:
- If there was a header for encoding, we try to get the encoding and its associated preamble length.
- If that wasn't successful, either because there wasn't a header or because we didn't recognize the encoding, we try to detect it from the data and get its associated BOM length
- Finally, if that wasn't successful, we use a default encoding and BOM length.